### PR TITLE
Provide testgrid.yaml support

### PR DESCRIPTION
**Pull request by kasunbg**
## Purpose
The document on '[Infrastructure / Deployment / and Scenarios Repository Structure - External consumers' view](https://github.com/wso2/testgrid/blob/95116ebae797d93f490d2ea3cd08a9b52b3df40c/docs/Infrastructure-Deployment-Scenarios-Repository-Structure.md)' details about introducing a configuration file each to describe Infrastructure, Deployment, and ScenarioTest (3) repositories. They are:

1. WSO2Infrastructurefile
2. WSO2Deploymentfile
3. WSO2Scenariosfile

The purpose of these files are detailed in the above document. But, that is not clearly distinguishable by the file name. It is too generic ATM. Therefore, after a round of discussion, we thought to finalize on following configuration file:

### testgrid.yaml

testgridyaml

The intention of this file is pretty clear - it is intended to be used with Testgrid. In here, we will namespace and add the three configurations originally discussed into this single config file, named testgrid.yaml. _Suggestions for a better name are welcome!_

The naming also goes in-line with industry practices. For example, Continuous Integration giant Jenkins require users to add a [Jenkinsfile](https://jenkins.io/doc/book/pipeline/jenkinsfile/) that describes how to build that particular repo (This was introduced in Jenkins2). Similarly, Travis CI needs users to place [.travis.yml](https://docs.travis-ci.com/user/customizing-the-build/) [config file](https://github.com/wso2/carbon-identity-framework/blob/master/.travis.yml) in their git repos to tell Travis CI what to do. `Testgrid.yaml` is there to do exactly that, but for Testgrid.

## Goals
1. Provide a way for infrastructure/deployment/scenario-test repositories to describe themselves, and tell Testgrid what to do with that repo.
2. Provide a convenient way for testgrid to know:
2.1 How to interpret the 'infrastructure' repos to provision infrastructure 
2.2 How to intepret the 'deployment' repos to do a WSO2 deployment in the provisioned infrastructure
2.3 How to interpret the 'scenario test' repos to execute scenario test scripts against the created WSO2 deployment.

## User stories
User stories for developers at each layer infrastructure/deployment/scenario-tests can be defined. This will be done as part of the above document.

## Documentation
[Infrastructure / Deployment / and Scenarios Repository Structure - External consumers' view](https://github.com/wso2/testgrid/blob/95116ebae797d93f490d2ea3cd08a9b52b3df40c/docs/Infrastructure-Deployment-Scenarios-Repository-Structure.md)

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
local-is-deployment and tomcat-deployment samples are updated with new format.

## Related PRs
N/A

## Migrations (if applicable)
Need to send a PR to cloudformation-is to use the new config file. @azinneera please note.

### CODE REVIEW

File: common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+package org.wso2.testgrid.common.config;
+
+import org.apache.commons.collections4.ListUtils;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * todo.
```
Comment (vidurananayakkara) [inline common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java]: Missed out java doc comment

File: common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        return ListUtils.emptyIfNull(deploymentPatterns);
+    }
+
+    public void setDeploymentPatterns(
+            List<DeploymentPattern> deploymentPatterns) {
+        this.deploymentPatterns = deploymentPatterns;
+    }
+
+    /**
+     * todo.
```
Comment (vidurananayakkara) [inline common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java]: Missed out java doc comment

File: common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
+/*
```
Comment (vidurananayakkara) [inline common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java]: All the public methods must include java doc comments. IMO this should be applicable for private methods as well considering future developers who will take over

File: common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
+/*
```
Comment (vidurananayakkara) [inline common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java]: All public methods must include java doc comments

File: common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        return ListUtils.emptyIfNull(provisioners);
+    }
+
+    public void setProvisioners(List<Provisioner> provisioners) {
+        this.provisioners = provisioners;
+    }
+
+    /**
+     * Provisioner has the same behavior as the {@link DeploymentConfig.DeploymentPattern}.
+     * todo
```
Comment (vidurananayakkara) [inline common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java]: Unnecessary todo in java doc

File: common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    /**
+     * Defines the infrastructure provider types.
+     *
+     * @since 1.0.0
+     */
+    public enum InfrastructureProvider {
+        AWS("AWS"),
+        OPENSTACK("OpenStack"),
+        GCP("GCP"),
+        SHELL("SHELL"),
```
Comment (sameerawickramasekara) [inline common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java]: Is SHELL appropriate here among other cloud providers?

File: common/src/main/java/org/wso2/testgrid/common/config/ScenarioConfig.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * denoted by the {@link TestgridYaml}.
+ *
+ */
+public class ScenarioConfig implements Serializable {
+    private static final long serialVersionUID = 6295205041044769906L;
+
+    private List<String> scenarios;
+
+    /**
+     *
```
Comment (sameerawickramasekara) [inline common/src/main/java/org/wso2/testgrid/common/config/ScenarioConfig.java]: Missing java doc subject line

File: common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java (hunk 6 above)
Comment (kasunbg) [reply common/src/main/java/org/wso2/testgrid/common/config/InfrastructureConfig.java]: No, we need to improve this logic. Created https://github.com/wso2/testgrid/issues/500 to track.

This was [existing logic](https://github.com/wso2/testgrid/pull/495/files/9682d888d42bf4ef2b50034c0a3edc9dd8f883cf#diff-83442d29c87f411acc105a9759934c6fL215) that was brought into a new class.

File: common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java (hunk 1 above)
Comment (kasunbg) [reply common/src/main/java/org/wso2/testgrid/common/config/DeploymentConfig.java]: Added all.

File: common/src/main/java/org/wso2/testgrid/common/config/Script.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        /**
+         * Defines the create-infra shell script type.
+         */
+        CREATE("Create"),
+
+        /**
+         * Defines the deploy shell script type.
+         */
+        DEPLOY("Deploy"),
+
```
Comment (azinneera) [inline common/src/main/java/org/wso2/testgrid/common/config/Script.java]: Don't we need an undeploy phase as well?

File: core/src/main/java/org/wso2/testgrid/core/command/CommandHandler.java [hunk 9]
```diff
@@ -103,14 +103,14 @@ public void execute() throws CommandExecutionException {
...
             final String ls = System.lineSeparator();
-            String versionBuilder = "WSO2 Deployment Monitor " + getVersion() + ls +
-                                    "Deployment Monitor Home: " +
-                                    System.getProperty("deployment.monitor.home", "<unknown>") + ls +
-                                    "Java version: " + System.getProperty("java.version", "<unknown>") + ls +
-                                    "Java home: " + System.getProperty("java.home", "<unknown>") + ls +
-                                    "OS name: \"" + SystemUtils.OS_NAME +
-                                    "\", version: \"" + SystemUtils.OS_VERSION +
-                                    "\", arch: \"" + SystemUtils.OS_ARCH + ls;
+            String versionBuilder = "WSO2 TestGrid " + getVersion() + ls +
```
Comment (azinneera) [inline core/src/main/java/org/wso2/testgrid/core/command/CommandHandler.java]: Shall we use StringUtil.concatStrings() here?

File: common/src/main/java/org/wso2/testgrid/common/config/Script.java (hunk 8 above)
Comment (kasunbg) [reply common/src/main/java/org/wso2/testgrid/common/config/Script.java]: Created https://github.com/wso2/testgrid/issues/501 to track.

**Pull request merged**